### PR TITLE
Implemented image upload on Chat

### LINKED ISSUE #1314: Chat control should support images
### Description

The chat control of #1092 is really cool! It would be a good addition to support images in the messages of the chat. This would be especially useful since now LLMs start supporting image vision and image generation. I could create cool demos around this chat control.

Florian:
People are also using chat demos to visualize tables inside their chat. This could also be a nice addition.

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by Satoshi-Sh**
## Related Issue 
resolves #1314

## Spec Implementation
In this PR, I implemented the feature to support images in the chat control. The images can now be sent and displayed alongside text messages within the chat. I followed the original spec, focusing on enabling the chat to render image URLs and display the corresponding images.

## Update 
- Add the attach file button to the message input 
- Add preview image with chip

## Todo 
- Add tests 
- Refactor the code for the formality. I will refer to `FileSelector.tsx` 
## Demo 

imge-upload-demo

I used the example from #1092. I added None to msgs as an image field.

```
from taipy.gui import Gui, Icon

msgs = [
    ["1", "msg 1", "Fred",None],
    ["2", "msg From Another unknown User", "Fredo",None],
    ["3", "This from the sender User", "taipy",None],
    ["4", "And from another known one", "Fredi",None],
]

users = [["Fred", Icon("/images/favicon.png", "Fred.png")], ["Fredi", Icon("/images/fred.png", "Fred.png")]]


def on_action(state, var_name: str, payload: dict):
    args = payload.get("args", [])
    msgs.append([f"{len(msgs) +1 }", args[2], args[3]])
    state.msgs = msgs

def button_action(state):
    msgs.loc[msgs.shape[0]] = [f"{msgs.shape[0] +1 }", "msg sent from Button", "taipy"]
    state.msgs = msgs

 
Gui(
    """
<|toggle|theme|>
# Test Chat
<|1 1 1|layout|
<|{msgs}|chat|users={users}|>

<|part|>

<|{msgs}|chat|users={users}|not with_input|>
|>

<|Send|button|on_action=button_action|>
""",
    path_mapping={"images": "c:\\users\\jeu\\downloads"},
).run()

```

### CODE REVIEW

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Good start
Could you write the spec you implemented in the issue ?
The image URL needs to be in a third column of the tabular data ?
Do you plan support for drag'n Drop and Copy/paste ?

**Comment by Satoshi-Sh** [comment]
Thanks for checking, Fred. 
I've added spec implementation to the first comment.

> The image URL needs to be in a third column of the tabular data ?
 
Yes, for this implementation. This is because of how `createSendActionNameAction`  is structured. Maybe I can modify the structure? 
 
> Do you plan support for drag'n Drop and Copy/paste ?

I'm not sure if they are necessary. I know they are implemented in the fileselector component. ChatGPT UI doesn't have either feature and Claude UI has only drag and drop. What do you think?

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx [hunk 1]
```diff
@@ -132,14 +133,15 @@ const defaultBoxSx = {
...
-
+const imageSx = {width:3/5, height:"auto"}
 interface key2Rows {
     key: string;
 }
 
 interface ChatRowProps {
     senderId: string;
     message: string;
+    image: string|null;
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.tsx]: could use `image?: string;`

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx [hunk 2]
```diff
@@ -424,6 +463,7 @@ const Chat = (props: ChatProps) => {
                                 senderId={senderId}
                                 message={`${row[columns[1]]}`}
                                 name={columns[2] ? `${row[columns[2]]}` : "Unknown"}
+                                image={row[columns[3]] ? `${row[columns[3]]}` : null}
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.tsx]: image={columns[3] && columns[3] != "_tp_index" ? `${row[columns[3]]}` : undefined}

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx [hunk 3]
```diff
@@ -256,30 +271,54 @@ const Chat = (props: ChatProps) => {
...
-        [updateVarName, onAction, senderId, id, dispatch, module]
+        [imagePreview,updateVarName, onAction, senderId, id, dispatch, module]
     );
 
+    const handleFileSelect = (event: React.ChangeEvent<HTMLInputElement>) => {
+            const file = event.target.files ? event.target.files[0] : null;
+            if (file) {
+                if (file.type.startsWith("image/")) {
+                    setSelectedFile(file);
+                    setImagePreview(URL.createObjectURL(file));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.tsx]: Should we limit file size ?

Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]: and invoke [URL.revokeObjectURL()](https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL_static) when the URL is not needed anymore ?

File: frontend/taipy-gui/src/components/Taipy/Chat.spec.tsx [hunk 4]
```diff
@@ -123,15 +123,61 @@ describe("Chat Component", () => {
...
+        const { getByLabelText,getByText,getByAltText,queryByText,getByRole } = render(
+            <TaipyContext.Provider value={{ state, dispatch }}>
+                <Chat messages={messages} updateVarName="varName" defaultKey={valueKey} mode="raw"/>
+            </TaipyContext.Provider>
+        );
+        const attachButton = getByLabelText('upload image');
+        expect(attachButton).toBeInTheDocument();
+
+        const file = new File(['(⌐□_□)'], 'test.png', { type: 'image/png' });
+        URL.createObjectURL = jest.fn(() => 'mocked-url');
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.spec.tsx]: create before test
and reset after test

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Looks very good, tx @Satoshi-Sh 
It would be very nice to add an example in doc/gui/examples

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx (hunk 2 above)
Comment (Satoshi-Sh) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]:
```
image={columns[3] && columns[3] != "_tp_index" ? `${row[columns[3]]}` : undefined}
```

I'm not sure how this logic works yet, but when I used this. The messages without image create `` element. 

Screenshot from 2024-10-18 08-59-35

I use this for now.
```
image={columns[3] && columns[3] != "_tp_index" && row[columns[3]] ? `${row[columns[3]]}` : undefined}
```

Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]: way better, tx

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx (hunk 3 above)
Comment (Satoshi-Sh) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]: The chat component needs to display the images with the URLs after message submissions. When should I revoke them? 

I added `max_file_size` property to the Chat component. I will add the test for this and consider adding a warning about file type and size.

File: frontend/taipy-gui/src/components/Taipy/Chat.spec.tsx (hunk 4 above)
Comment (Satoshi-Sh) [reply frontend/taipy-gui/src/components/Taipy/Chat.spec.tsx]: I moved
```
const file = new File(['(⌐□_□)'], 'test.png', { type: 'image/png' });
URL.createObjectURL = jest.fn(() => 'mocked-url');
```
before the `expect(attachButton).toBeInTheDocument()`;

and add jest.restoreAllMocks() at the end.  Let me know if that's not what you mean.

**Comment by Satoshi-Sh** [comment]
## Examples Update 
I updated the existing chat example files and created a new example file from Fred's example. 

A new example script is like this image.
Screenshot from 2024-10-18 13-44-46

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx [hunk 5]
```diff
@@ -34,7 +34,7 @@ import Paper from "@mui/material/Paper";
 import Popper from "@mui/material/Popper";
 import TextField from "@mui/material/TextField";
 import Tooltip from "@mui/material/Tooltip";
-import Send from "@mui/icons-material/Send";
+import {Send, AttachFile} from "@mui/icons-material";
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.tsx]: Coding rules: separate imports for better tree shaking

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx (hunk 3 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]: When the component is unmounted?
In the returned function from use effect...

File: taipy/gui/_renderers/factory.py [hunk 6]
```diff
@@ -99,6 +99,7 @@ class _Factory:
                 ("sender_id",),
                 ("height",),
                 ("page_size", PropertyType.number, 50),
+                ("max_file_size", PropertyType.number, 50 * 1024 * 1024),
```
Comment (FredLL-Avaiga) [inline taipy/gui/_renderers/factory.py]: 50MB seems a lot?
I'd say 1MB seems right.

**Comment by FredLL-Avaiga** [comment]
> ## Examples Update 
> I updated the existing chat example files and created a new example file from Fred's example. 
> 
> A new example script is like this image.
> Screenshot from 2024-10-18 13-44-46
> 

How does it render with show_sender=True?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

**Comment by Satoshi-Sh** [comment]
> How does it render with show_sender=True?

I just updated the example script with show_sender.

Screenshot from 2024-10-21 16-44-24

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx (hunk 3 above)
Comment (Satoshi-Sh) [reply frontend/taipy-gui/src/components/Taipy/Chat.tsx]: I tried to implement but not sure if it's right. Please have a look at it.

**Comment by FredLL-Avaiga** [comment]
> > How does it render with show_sender=True?
> 
> I just updated the example script with show_sender.
> 
Not sure about this 
How do other chats behave in this situation ?

File: frontend/taipy-gui/src/components/Taipy/Chat.tsx [hunk 7]
```diff
@@ -390,6 +435,14 @@ const Chat = (props: ChatProps) => {
         loadMoreItems(0);
     }, [loadMoreItems]);
 
+    useEffect(() => {
+        return () => {
+            for (const objectURL of objectURLs) {
+                URL.revokeObjectURL(objectURL);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chat.tsx:441]: will have to be mocked for tests I suppose

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
test failing

**Comment by Satoshi-Sh** [comment]
> Not sure about this
How do other chats behave in this situation ?

I thought you wanted to see what looked like with show_sender=True. That is why I updated `chat_messages.py` with show_sender=True.  I'm not sure what you expect here.

**Comment by FredLL-Avaiga** [comment]
> > Not sure about this
> > How do other chats behave in this situation ?
> 
> I thought you wanted to see what looked like with show_sender=True. That is why I updated `chat_messages.py` with show_sender=True. I'm not sure what you expect here.

Thank you for showing me what it looks like
I must say I don't think it look very good, that's why I wonder how chat apps (whatsapp, Teams ...) handle this kind of display

**Comment by Satoshi-Sh** [comment]
Thanks for the clarification. 

I just checked other chat apps. It seems uncommon to display the sender's avatar and name in the message. ChatGPT separate image and message. Whatsapp puts an image in the same speech bubble.

## chatgpt
Screenshot from 2024-10-22 08-15-48

## whatsapp
Screenshot from 2024-10-22 08-24-02

**Comment by FredLL-Avaiga** [comment]
@jrobinAV I think you can merge this PR

**Review by FredLL-Avaiga — APPROVED** [review]

**Comment by jrobinAV** [comment]
Thank you @Satoshi-Sh

**Comment by Satoshi-Sh** [comment]
Thanks for the guidance @FredLL-Avaiga.

**Comment by RymMichaut** [comment]
Hey @Satoshi-Sh 
Congratulations on this merged PR.
Could you please contact me at [rym.michaut@taipy.io](mailto:rym.michaut@taipy.io) to prepare and send your swag?

**Comment by Satoshi-Sh** [comment]
@quest-bot loot #1314

**Pull request merged**